### PR TITLE
fix(rerank): avoid manufactured top-confidence on uniformly-weak out-of-range batches

### LINKED ISSUE #16621: [Bug]: Issues with Min-Max Scaling for Reranker Output
### Self Checks

- [x] I have searched for existing issues [search for existing issues](https://github.com/infiniflow/ragflow/issues), including closed ones.
- [x] I confirm that I am using English to submit this report ([Language Policy](https://github.com/infiniflow/ragflow/issues/5910)).
- [x] Non-english title submitions will be closed directly ( 非英文标题的提交将会被直接关闭 ) ([Language Policy](https://github.com/infiniflow/ragflow/issues/5910)).
- [x] Please do not modify this template :) and fill in all the required fields.

### RAGFlow workspace code commit ID

main

### RAGFlow image version

main

### Other environment information

```Markdown
It is not ideal to apply Min-Max normalization to reranker scores, as it forces the maximum score to 1. For example, given reranked scores like [0.3, 0.2, 0.1], where none indicate high similarity, Min-Max normalization would set the first to 1.0. This artificially inflates the final score when combined with term-based similarity via weighted fusion.
```

### Actual behavior

It is not ideal to apply Min-Max normalization to reranker scores, as it forces the maximum score to 1. For example, given reranked scores like [0.3, 0.2, 0.1], where none indicate high similarity, Min-Max normalization would set the first to 1.0. This artificially inflates the final score when combined with term-based similarity via weighted fusion.

### Expected behavior

_No response_

### Steps to reproduce

```Markdown
run retrieve() with reranker model.
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Harsh23Kashyap**
### What problem does this PR solve?

Issue [#16621](https://github.com/infiniflow/ragflow/issues/16621) describes Min-Max normalization producing inflated top scores when the reranker's raw output is out of `[0, 1]`. The example in the issue body (`[0.3, 0.2, 0.1]` → alleged `[1.0, 0.66, 0.33]`) is already preserved verbatim by the current `Base._normalize_rank` (the in-range short-circuit fires first), but the failure signature is real on the out-of-range branch: a uniformly-weak batch like `[-1.0, -1.5, -2.0]` got min-max stretched to `[1.0, 0.5, 0.0]`. With `vtweight=0.7` in `rag/nlp/search.py`'s `tkweight * tksim + vtweight * vtsim` hybrid blend, the top chunk then dominated the ranking despite no chunk actually matching strongly — manufactured top confidence.

### What changed?

`rag/llm/rerank_model.py::Base._normalize_rank` now uses a centred sigmoid on the out-of-range branch: shift each score to the batch midpoint, then squash to `(0, 1)`. Centring keeps absolute confidence meaningful — a uniformly-weak batch lands around `[0.62, 0.50, 0.38]` rather than `[1.0, 0.5, 0.0]`, and a tight cluster well above zero lands firmly above `0.5` to reflect real absolute confidence. Sigmoid is monotone so ordering within the batch is preserved.

### Backward compatibility

- In-range providers (Cohere / Jina / Voyage) byte-identical: the `min_rank >= 0 and max_rank <= 1` short-circuit still fires before any rescaling.
- Out-of-range providers (NVIDIA) get smoother scores around `0.5` instead of the previous forced stretch to `[0, 1]`.
- API surface unchanged: `similarity()` still returns `(rank, token_count)`.

### Testing

- Existing `test_out_of_range_scores_are_rescaled` and `test_nvidia_logits_are_normalized` updated to expect centred-sigmoid outputs.
- New `test_uniformly_weak_out_of_range_does_not_manufacture_top_confidence`: parametrised over uniformly-weak batches, asserts top score is bounded (no longer forced to `1.0`) and intra-batch ordering is preserved.
- New `test_centred_sigmoid_centers_at_batch_mean`: asserts the midpoint of an out-of-range batch maps to exactly `0.5` and the result is strictly monotone.

All 22 tests in `test/unit_test/rag/llm/test_rerank_normalization.py` pass.

```
$ uv run pytest test/unit_test/rag/llm/test_rerank_normalization.py -q
22 passed in 0.30s
```

### Files changed

- `rag/llm/rerank_model.py` (1 method, +14 / −6)
- `test/unit_test/rag/llm/test_rerank_normalization.py` (+67 / −13)

Fixes #16621

### CODE REVIEW

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

File: rag/llm/rerank_model.py [hunk 1]
```diff
@@ -86,10 +98,12 @@ def _normalize_rank(rank: np.ndarray) -> np.ndarray:
...
             return rank
-        span = max_rank - min_rank
-        if span < 1e-3:
+        if max_rank - min_rank < 1e-3:
             return np.clip(rank, 0.0, 1.0)
-        return (rank - min_rank) / span
+        # Centre on the batch midpoint, then squash. Min-max stretch was
+        # replaced because it produced [1.0, ...] for uniformly-weak batches,
+        # which manufactured false top-confidence under vtweight=0.7.
+        return 1.0 / (1.0 + np.exp(-(rank - (min_rank + max_rank) / 2.0)))
```
[the remark this answers is not in the document]
Comment (Harsh23Kashyap) [reply rag/llm/rerank_model.py:107]: Pushed `25849c707` addressing this. The sigmoid offset is now divided by the batch span, so wide uniformly-weak batches (e.g. [-10, -20, -30]) and small-spread equivalents (e.g. [-1, -1.5, -2]) collapse to the same scores near 0.5 rather than producing near-1.0 confidence.

Added a new test `test_centred_sigmoid_invariant_to_absolute_batch_magnitude` to pin that invariance, plus a `[-10, -20, -30]` case in the existing false-confidence parametrisation. Full pytest run: 23 passed.

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
Can someone take a look when you have a chance?

**Comment by Harsh23Kashyap** [comment]
CodeRabbit thread resolved in 25849c707 (sigmoid offset now divided by batch span). CI green.

**Comment by Harsh23Kashyap** [comment]
CodeRabbit thread resolved in 25849c707 (sigmoid offset now divided by batch span). CI green.

**Comment by Harsh23Kashyap** [comment]
CodeRabbit thread resolved in 25849c707 (sigmoid offset now divided by batch span). Standing by for a maintainer pass.
